core/src/main/java/org/infinispan/container/entries/DeltaAwareCacheEntry.java

```diff
@@ -146,16 +148,26 @@ public final K getKey() {
 
    @Override
    public final DeltaAware getValue() {
-      return value;
+      // TODO: it's not possible to return the actually modified (but uncommitted) value
+      // when this is not a copyable entry
+      return uncommittedChanges;
    }
 
    @Override
    public final DeltaAware setValue(DeltaAware value) {
-      DeltaAware oldValue = this.value;
-      this.value = (DeltaAware) value;
+      DeltaAware oldValue = getValue();
+      this.value = value;
+      // TODO: this won't work without the copy for normal deltas
```

Comment:
I think that the problem is that when the value is not AtomicHashMap, we don't copy it and modify the original DeltaAware value.